rust/processor/src/models/token_models/ans_lookup.rs

```diff
@@ -74,107 +73,396 @@ impl OptionalString {
     }
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct NameRecordKeyV1 {
+    domain_name: String,
+    subdomain_name: OptionalString,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct NameRecordV1 {
+    #[serde(deserialize_with = "deserialize_from_string")]
+    expiration_time_sec: BigDecimal,
+    #[serde(deserialize_with = "deserialize_from_string")]
+    property_version: BigDecimal,
+    target_address: OptionalString,
+}
+
 impl CurrentAnsLookup {
     pub fn from_transaction(
         transaction: &TransactionPB,
-        ans_contract_address: Option<String>,
+        address_to_primary_name_record: &mut AddressToPrimaryNameRecord,
+        maybe_ans_primary_names_table_handle: Option<String>,
+        maybe_ans_name_records_table_handle: Option<String>,
+        conn: &mut PgPoolConnection,
     ) -> HashMap<CurrentAnsLookupPK, Self> {
         let mut current_ans_lookups: HashMap<CurrentAnsLookupPK, Self> = HashMap::new();
-        if let Some(addr) = ans_contract_address {
-            // Extracts events and user request from genesis and user transactions. Other transactions won't have coin events
-            let txn_data = transaction
-                .txn_data
+        let txn_version = transaction.version as i64;
+        let txn_data = transaction
+            .txn_data
+            .as_ref()
+            .expect("Txn Data doesn't exit!");
+
+        // Extracts from user transactions. Other transactions won't have any ANS changes
+        if let (
+            TxnData::User(_),
+            Some(ans_primary_names_table_handle),
+            Some(ans_name_records_table_handle),
+        ) = (
+            txn_data,
+            maybe_ans_primary_names_table_handle,
+            maybe_ans_name_records_table_handle,
+        ) {
+            let transaction_info = transaction
+                .info
                 .as_ref()
-                .expect("Txn Data doesn't exit!");
-            if let TxnData::User(user_txn) = txn_data {
-                for event in &user_txn.events {
-                    let (event_addr, event_type) = if let Content::Struct(inner) =
-                        event.r#type.as_ref().unwrap().content.as_ref().unwrap()
-                    {
-                        (
-                            inner.address.to_string(),
-                            format!("{}::{}", inner.module, inner.name),
+                .expect("Transaction info doesn't exist!");
+
+            // Extract primary names and name records
+            for wsc in &transaction_info.changes {
+                if let Ok(name_records) = match wsc.change.as_ref().unwrap() {
```

Comment:
Hmm change shouldn't be a Result. Isn't it an optional and .unwrap should already expose the struct?

Reply:
It's unwrapping the Result returned by `Self::from_write_table_item`